Mod Loading Support

Mr-Wiseguy opened the pull request:
## Overview
This is a PR of the in-development `modding` branch. This branch represents the work done so far for the modding system that will be available in 1.2.

It currently supports the following features:
* Loading N64Recomp mods (`.nrm` files), which can edit game code or add new code to the game.
* Loading RT64 texture packs (`.rtz` files).
  * Texture packs can also be embedded into `.nrm` files as well.
> [!NOTE]
> For anyone who was using the previous texture pack PR (#447), the folder for mods has been changed from `mods/mm` to just `mods`.

The following mod functionality will be added in future PRs:
* A mod management menu to browse installed mods and turn them off or on.
* The ability for mods to include their own custom menus and config options.
* Bundling a few built-in mods directly with Zelda 64: Recompiled (exact set of built-in mods to be determined).

## Instructions
1. Download a build for your platform from the next section. ⬇️
2. Create the mod folder as specified in the **Mods Directory** section.
3. Download a mod in `.nrm` or `.rtz` format and place it in the folder created in step 2. Also place any `.dll`/`.so` files that came with the mod in that same folder.

<!--- section:artifacts:start -->
### Build Artifacts
- [Zelda64Recompiled-Windows-RelWithDebInfo.zip](https://nightly.link/Zelda64Recomp/Zelda64Recomp/actions/artifacts/2590457498.zip)
- [Zelda64Recompiled-Linux-X64-Release.zip](https://nightly.link/Zelda64Recomp/Zelda64Recomp/actions/artifacts/2590440548.zip)
- [Zelda64Recompiled-AppImage-X64-Release.zip](https://nightly.link/Zelda64Recomp/Zelda64Recomp/actions/artifacts/2590441171.zip)
- [Zelda64Recompiled-Linux-ARM64-Release.zip](https://nightly.link/Zelda64Recomp/Zelda64Recomp/actions/artifacts/2590440237.zip)
- [Zelda64Recompiled-AppImage-ARM64-Release.zip](https://nightly.link/Zelda64Recomp/Zelda64Recomp/actions/artifacts/2590442262.zip)

### Debug Builds for Developers
- [Zelda64Recompiled-Windows-Debug.zip](https://nightly.link/Zelda64Recomp/Zelda64Recomp/actions/artifacts/2590440459.zip)
- [Zelda64Recompiled-Linux-X64-Debug.zip](https://nightly.link/Zelda64Recomp/Zelda64Recomp/actions/artifacts/2590440685.zip)
- [Zelda64Recompiled-AppImage-X64-Debug.zip](https://nightly.link/Zelda64Recomp/Zelda64Recomp/actions/artifacts/2590441367.zip)
- [Zelda64Recompiled-Linux-ARM64-Debug.zip](https://nightly.link/Zelda64Recomp/Zelda64Recomp/actions/artifacts/2590439883.zip)
- [Zelda64Recompiled-AppImage-ARM64-Debug.zip](https://nightly.link/Zelda64Recomp/Zelda64Recomp/actions/artifacts/2590441734.zip)
- [Zelda64Recompiled-PDB-Debug.zip](https://nightly.link/Zelda64Recomp/Zelda64Recomp/actions/artifacts/2590440627.zip)
- [Zelda64Recompiled-PDB-RelWithDebInfo.zip](https://nightly.link/Zelda64Recomp/Zelda64Recomp/actions/artifacts/2590457689.zip)
<!--- section:artifacts:end -->

## Mods Directory
You'll need to create the following folder yourself and move any mods you download into it:
- **Windows**: `%LocalAppData%/Zelda64Recompiled/mods`
- **Linux**: `~/.config/Zelda64Recompiled/mods`

## Currently Available Texture Packs

### Majora's Mask N64HD Project

HD textures drawn by a sultry-lipped CGI moon monster to match the original art.

* **Author**: Nerrel
* **Pack Download (.rtz)**: https://github.com/Nerrel/MMHD/releases/latest
* **Website**: https://nerrel.com/mmhd/
* **Additional contributors**:  GhostlyDark, Tom Garden

### Majora's Mask Reloaded

Majora's Mask Reloaded is a custom version of Nerrel's high quality MMHD texture pack. It features texture edits that closer resemble the N64 original visuals, as well as improved language and ROM hack support.
* **Pack Download (.rtz)**: https://github.com/GhostlyDark/MM-Reloaded/releases/tag/2024-07-28
* **Website**: https://evilgames.eu/texture-packs/mm-reloaded.htm
* **Authors**: GhostlyDark, Nerrel
* **Additional contributors (cross platform)**: Admentus, Asra, BLUEamnesiac, Goodland, Darth-Koopa, OoT Community Retexture Team, Pizza, tao, Third M, Tjk918, yonato

## Details For Mod Developers
For more details on what N64: Recompiled mods can do and how to make them, check out [this documentation on modding](https://hackmd.io/fMDiGEJ9TBSjomuZZOgzNg?view) and the [Majora's Mask: Recompiled mod template](https://github.com/Zelda64Recomp/MMRecompModTemplate).
As 1.2 gets closer to release a more in-depth guide will be available for writing mods, but the two linked resources should be enough to get started with making simple mods.

Mr-Wiseguy committed:
Updated RT64 and enabled RT64 extended RDRAM mode

Mr-Wiseguy committed:
Removed shader cache

Mr-Wiseguy committed:
Update modern runtime for mod support, add some exports and events in the patches

Mr-Wiseguy committed:
Update N64Recomp commit for CI

Mr-Wiseguy committed:
Parse version number from a string and provide it to the runtime

Mr-Wiseguy committed:
Update runtime for new mod manifest schema

Mr-Wiseguy committed:
Update runtime to have mod list functionality

Mr-Wiseguy committed:
Update runtime and print mod authors and dependencies

Mr-Wiseguy committed:
Update runtime for mod loading on posix systems, enable all mods temporarily

Mr-Wiseguy committed:
Update N64ModernRuntime for mod mips32 relocs and function lookups

Mr-Wiseguy committed:
Update runtime after merge

Mr-Wiseguy committed:
Update runtime to fix reference symbols

Mr-Wiseguy committed:
Update runtime for populating missing runtime pointers in offline compiled mods

Mr-Wiseguy committed:
Update runtime after merge

Mr-Wiseguy committed:
Update runtime to add DLL search path, enable strict mode in patch recompilation

Mr-Wiseguy committed:
Update runtime for custom mod content types and add RT64 texture pack support

LittleCube-hax committed:
add exports and events for moon crash save and owl save (#487)

* add exports and events for moon crash save and owl save

* fix recomp_on_owl_save event, add exports and callbacks for save loading

* add more flexible owl events

* add init event

* fix init event to not be terrible

* rename a couple events

* use deletes instead of resets

* use better names, add better annotations

* use full signature for event annotations

Mr-Wiseguy committed:
Disable identical code folding to prevent mods from conflicting with themselves due to merged functions

dcvz committed:
chore(ci): Bundle Windows PDBs  (#492)

* update runners

* Save PDB

* Build RelWithDebInfo

* Update copy on runners

* Archive PDB separately

* Update copy

* Update PDB archive name

Mr-Wiseguy committed:
Add custom symbols toml and made non-relocatable versions of original gamestate functions, added base patch for Play_Init

Mr-Wiseguy committed:
Update runtime to fix mod section alignment

LittleCube-hax committed:
Add events and exports for Play_Init (#495)

LittleCube-hax committed:
Add exports for fd anywhere and epona fix patches (#494)

* add exports for fd anywhere and epona fix patches

* use better logic for restoreHudVisibility

* cleanup

* fix a couple edge cases (being sent the bow while riding Epona)

LittleCube-hax committed:
Add recomp_after_play_init event (#496)

* fix signature for recomp_set_allow_no_ocarina_tf, add new event recomp_after_play_init

* say initialize instead of construct

LittleCube-hax committed:
More vanilla bug fix exports (#497)

Mr-Wiseguy committed:
Update runtime for heap allocator

LT-Schmiddy commented:
Question, do these builds support calling functions from external libraries? I can see information related to that in the my `mod.toml`, but I can't figure out how to actually call 'em. If it IS currently possible, can someone walk me through it?

I'm working on a collection of tweaks and QoL improvements in a mod. I'd like to be able to have a config file, but I need to be able to build a separate library for that to work.

(I know Zelda64Recomp may include something like this built-in in the future, but I'm impatient ;P )

Mr-Wiseguy commented:
The current builds do support calling external libraries. There's a mod being worked on that does that you can reference as an example. The relevant section of the toml for that mod is here: <https://github.com/LittleCube-hax/MMRecompRando/blob/58e7b58a464fc71beb4294badeebb37268457151/mod.toml#L28> and the implementations of those functions are here: <https://github.com/LittleCube-hax/MMRecompAPCppGlue/blob/fced9ab75e77398858bd936c1cf62cfaa86f13db/apcpp-glue.cpp#L82>. You'll also need that recomp_api_version export.

With that said mods will have the ability to add their own custom menu to the config UI by providing a json file that describes the mod's options, so you don't actually need to have custom code for reading your own config file. I'll give an update here when that's done; it's what I'm working on currently.

DarioSamo commented:
A user has reported that the appimage of this build crashes when trying to open the "Select ROM" file dialogue, while the appimage for 1.1.1 works fine. Was there any change regarding this between both builds?

Mr-Wiseguy committed:
Update RT64 for optimized ubershaders and update recomp runtime

LT-Schmiddy commented:
> The current builds do support calling external libraries. There's a mod being worked on that does that you can reference as an example. The relevant section of the toml for that mod is here: https://github.com/LittleCube-hax/MMRecompRando/blob/58e7b58a464fc71beb4294badeebb37268457151/mod.toml#L28 and the implementations of those functions are here: https://github.com/LittleCube-hax/MMRecompAPCppGlue/blob/fced9ab75e77398858bd936c1cf62cfaa86f13db/apcpp-glue.cpp#L82. You'll also need that recomp_api_version export.
> 
> With that said mods will have the ability to add their own custom menu to the config UI by providing a json file that describes the mod's options, so you don't actually need to have custom code for reading your own config file. I'll give an update here when that's done; it's what I'm working on currently.

Ok, thanks! I've looked through the code you linked, and I think I see how it works.

In retrospect, I'll probably just split up the tweaks I'm developing into multiple mods, so as to minimize any compatibility issues with other mods re-writing the same functions as I am. 

The first thing I'm doing is a Permanent Razor Sword patch. I've figured out how to prevent the sword from being reverted by studying 2Ship's implementation of the same feature, and it works fine. I'm still trying to figure out how to change the Mountain Smith's behavior and dialog to reflect this, however. 

Still a work in progress, but I've published my source-code [here](https://github.com/LT-Schmiddy/MMPremanentRazorSwordMod).

a-priestley commented:
Hello!

For anyone on Arch Linux, please have a look at my AUR submission for this draft [here](https://aur.archlinux.org/packages/zelda64recomp-modding-bin).

It targets both the x86_64 and aarch64 build artifacts linked at the top of this thread.

Cheers!

Mr-Wiseguy committed:
Update RT64 for S2DEX replacements

danielryb committed:
Add API function export for camera Z-targeting fixes (#503)

LT-Schmiddy committed:
Added Recomp Events for before and after first-person aiming update. (#511)

* Added Recomp Events for first-person aiming.

* Mods can now force right-stick aiming.

* Refactored aiming event code for formatting and clarity.

DarioSamo committed:
Use SDL Window as the WindowHandle (Enables wayland support). (#520)

DarioSamo committed:
Update RT64 to fix SDL2 Deadlock issues with ImGui. (#522)

Mr-Wiseguy committed:
Update runtime to integrate live recompiler (#524)

Mr-Wiseguy committed:
Update runtime for updated sse2neon, remove top level sse2neon submodule

Mr-Wiseguy committed:
Modding function hooks (#530)

This commit updates the runtime for function hooking and implements any recomp-side functionality needed for them to fully hook (namely ROM decompression). For more details, see the relevant N64Recomp and N64ModernRuntime PRs.

Mr-Wiseguy committed:
Update runtime for mod manifest changes

Mr-Wiseguy committed:
Update N64Recomp CI commit and runtime for patch regeneration fix

DarioSamo committed:
Update upload-artifact version.

Mr-Wiseguy committed:
Update runtime to fix hooking issue with base patched functions

Mr-Wiseguy committed:
Update runtime to fix jump tables in hooked functions

Mr-Wiseguy committed:
Unify API function naming scheme and export relevant API functions
